### DISCUSSION TITLE
timers and strange behaviour

### DISCUSSION THREAD

**Discussion by DRBAUS** [RomanKubiak/ctrlr#442] [Q&A]
Hi all,

I have the following (simplified) script working and running.

`--
-- Called when a double-click event occurs on this component
--

L1E1FULLON = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)

timer:setCallback(0, TimerCallback) -- TimerId 0
timer:startTimer(0, 500) -- Timer Id 0 500ms

end

counter1 = 1

TimerCallback = function(timerId00)

--TIME 1--

	if counter1 == 1 then
		counter1 = 2

		panel:sendMidiMessageNow(CtrlrMidiMessage({0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x04, 0x03, 0x00, 0x00, 0x00, 0x14, 0xF7}))

--TIME 2--
		
elseif counter1 == 2 then

		counter1 = 1	

		panel:sendMidiMessageNow(CtrlrMidiMessage({0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x04, 0x13, 0x00, 0x00, 0x00, 0x14, 0xF7}))

		timer:stopTimer(0)
end
end`

However when I copy this script and change the timer to for example Id 1 and run it either from the same double click on label or even from a double click on a different label it starts to behave unpredictable. Usually after saving if I double click each of the label it starts sending the sysex from the other script.

Any suggestion what might cause this?

**Comment by damiensellier** [comment]
You have a loop running. The timer 0 never stops since you set counter1  back to 1 when you have set it to 2.

SInce you have the same sysex message in both timers I don't understand your script.
If you want to send the same message twice on double click with a delay of 500ms just do counter1 = counter1 + 1
with a statement like:


```
L1E1FULLON = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)

counter1 = 0
timer:setCallback(0, TimerCallback) -- TimerId 0
timer:startTimer(0, 500) -- Timer Id 0 500ms

end
```


```
TimerCallback = function(timerId)

if timerId == 0 then

	if counter1 < 3 then
console("counter1 : "..(counter1))
		counter1 = counter1 + 1
		panel:sendMidiMessageNow(CtrlrMidiMessage({0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x04, 0x03, 0x00, 0x00, 0x00, 0x14, 0xF7}))

	elseif counter1 == 3 then
		timer:stopTimer(0)
	end

end -- end timer #0
```


end -- f



[SendSysexTwiceOnDbleClick_1_0_0_2022-03-07_04-52.panel.zip](https://github.com/RomanKubiak/ctrlr/files/8196628/SendSysexTwiceOnDbleClick_1_0_0_2022-03-07_04-52.panel.zip)


Can you please explain precisely what you want to achieve?

**Reply by DRBAUS** [reply]
The sysex messages are different. Message 1 is for ENV1 Time 1, Message 2 is for ENV1 Time 2.

I need to send 15 sysex messages (parameter values) with an interval of at least 200ms to my Ensoniq sampler. This is one of the Envelope Presets of which the sampler has multiple.

**Reply by damiensellier** [reply]
Oh okay it's almost the same.
```

TimerCallback = function(timerId)

if timerId == 0 then

	if counter1 == 1 then
		console("counter1 : "..(counter1))
		counter1 = counter1 + 1
		panel:sendMidiMessageNow(CtrlrMidiMessage({0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x04, 0x03, 0x00, 0x00, 0x00, 0x14, 0xF7}))
	
	elseif counter1 == 2 then
		console("counter1 : "..(counter1))
		counter1 = counter1 + 1
		panel:sendMidiMessageNow(CtrlrMidiMessage({0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x04, 0x13, 0x00, 0x00, 0x00, 0x14, 0xF7}))

-- end so on... add as much delayed messages as you want, for ex 15

	elseif counter1 == 16 then  -- when all the 15 messages have been sent counter1 reaches 16 (15+1) and stops the timer
		timer:stopTimer(0)
	end

end -- end timer #0


end -- f
```

if you have 15 messages the counter needs to go to counter1 == 16 to stop the timer. Don't forget to increment the counter1 +1 for every message like I did..

I use this kind of timer all the time for exemple with the ensoniq button press and release that requires a delay of at least 200ms between messages.

Let me know if it's working
cheers

**Reply by DRBAUS** [reply]
Thanks Damien!

It works...just once. 1x Double click and it sends the code. The next time double click nothing happens. Neither on the same label for the script containing timer 0 nor for a double click on the other label for the script containing timer 1.

**Reply by damiensellier** [reply]
You have 2 timers triggered by 2 different scripts, 1 for single click, a 2nd one for double click, that's it?

are you sure you set your counter1 value ?

L1E1FULLON = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)

**counter1 = 1**
timer:setCallback(0, TimerCallback) -- TimerId 0
timer:startTimer(0, 500) -- Timer Id 0 500ms

end

please check bcause I tried and it wrks like a charm.
Don't forget to put some terminal message to follow where you are at in the lua monitor. when debugging.


EDIT new panel uploaded

[SendSysexTwiceOnDbleClick_1_0_1_2022-03-07_07-44.panel.zip](https://github.com/RomanKubiak/ctrlr/files/8197802/SendSysexTwiceOnDbleClick_1_0_1_2022-03-07_07-44.panel.zip)

**Reply by damiensellier** [reply]
> Thanks Damien!
> 
> It works...just once. 1x Double click and it sends the code. The next time double click nothing happens. Neither on the same label for the script containing timer 0 nor for a double click on the other label for the script containing timer 1.

Please try this panel

https://github.com/RomanKubiak/ctrlr/files/8197802/SendSysexTwiceOnDbleClick_1_0_1_2022-03-07_07-44.panel.zip

![working on single and double click](https://user-images.githubusercontent.com/55164385/157039891-15b0b4b4-0d6d-450d-8018-d8b8e12e387a.gif)


PS: if you want to use 2 different timers on two different events you have to trigger the appropriate timer via its ID

counter1 = 1
timer:setCallback(**0**, TimerCallback) -- TimerId **0**
timer:startTimer(**0**, 500) -- Timer Id **0** 500ms


for the second timer 

counter2 = 1
timer:setCallback(**1**, TimerCallback) -- TimerId **1**
timer:startTimer(**1**, 500) -- Timer Id **1** 500ms

...

**Reply by dobo365** [reply]
@damiensellier The code you proposed with the counter is exactly what I meant as answer to another post ;-)
Just didn't have the time to write it fully and test it...

**Reply by damiensellier** [reply]
I don't know if it worked for @DRBAUS but this is actually the only way to do it.
You can streamline the process with a table and iteration to remove everything redundant lines if ever there's just 1 byte in the message changing for the different parameters (1 to 16) to reset to 0.  If they are not the same number as the counter number, we can crossmatch it with a lookuptable and use counter as "k"  to a table in pair for the corresponding modulator value.


```
TimerCallback = function(timerId)

if timerId == 0 then

	 if counter1 < 16 then
		console("counter1 : "..(counter1))
		panel:sendMidiMessageNow(CtrlrMidiMessage({0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x04, 0x03, 0x00, 0x00, 0x00, counter1, 0xF7}))
		counter1 = counter1 + 1
	else
		timer:stopTimer(0)
	end

end -- end timer #0

end -- f

```


with the Lookup table that will do something like  that : 




```
TimerCallback = function(timerId)


if timerId == 0 then


	 if counter1 < 16 then
		console("counter1 : "..(counter1))

		local tableMod = {2,6,3,18,27,4,3,6,8,6,1,5,9,15,14,21}

		for k,v in pairs(tableMod) do
			tableModValue = tableMod[counter1]
  		end  
            		console("tablemod value : "..tableModValue)
			panel:sendMidiMessageNow(CtrlrMidiMessage(MemoryBlock({0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x04, 0x03, 0x00, 0x00, 0x00, tableModValue, 0xF7})))
	
		counter1 = counter1 + 1

	else

		timer:stopTimer(0)
	end

end -- end timer #0

end -- f


```

**Reply by dobo365** [reply]
Lovely! Great Damien!

**Reply by DRBAUS** [reply]
Okay, could you please check the attached panel?

Your provided code is working! I changed the Id of the Timer to 0 and the name of the counter to counter0. Still works.

Then I copied the uiLabel, created another script called myMethod2, copied the code and changed the name counter0 to counter1 as well as timer Id 0 to Id1. I did the same for the TimerCallback. Created a TimerCallback1, copied the code from TimerCallback and changed counter0 to counter1 and timerId from 0 to 1.

Double click on that one only shows the following in console: 
>>LUA timer starts on double click
>>LUA timer starts on double click and counter1=1

and the rest of the script is not running.
[SendSysexOnDbleClick_1_0_1_2022-03-08_19-03.zip](https://github.com/RomanKubiak/ctrlr/files/8208683/SendSysexOnDbleClick_1_0_1_2022-03-08_19-03.zip)

**Reply by damiensellier** [reply]
@DRBAUS you need just one TimerCallback()  method for your panel. just call the different timers by their ids.
yours was not working because on your second method you called the timer id 1 in the TimerCallback() script that does not exist but is in TimerCallback1()

it would have been :

```
timer:setCallback(1, TimerCallback1) -- TimerId 1
timer:startTimer(1, 500) -- Timer Id 1 500ms

```
just put all your timers in the same method or you will get lost.

check the following panel I revised it.

[SendSysexTwiceOnDbleClick_1_1_0_2022-03-08_14-19.panel.zip](https://github.com/RomanKubiak/ctrlr/files/8209019/SendSysexTwiceOnDbleClick_1_1_0_2022-03-08_14-19.panel.zip)

 Also check the streamlined version of the panel I posted yesterday using tables to remove those 32 redundant "if". 
 https://github.com/RomanKubiak/ctrlr/discussions/442#discussioncomment-2312033

**Comment by DRBAUS** [comment]
Thanks guys.

I am have not been able to test it further today because of work. Will do
as soon as possible.

I think I noticed where it went wrong though.

Will update as soon as I can.

Op ma 7 mrt. 2022 22:05 schreef Damien S. ***@***.***>:

**Reply by damiensellier** [reply]
![send sysex with delay between messages](https://user-images.githubusercontent.com/55164385/157125402-014d662b-0940-426f-95e6-c8fbf17801fb.gif)



Here is a working panel with lookup table to the modulators number relative to the counter as "k" in table,  if needed:
[SendSysexWithDelayInbetween_1_0_3_2022-03-07_16-57.panel.zip](https://github.com/RomanKubiak/ctrlr/files/8201141/SendSysexWithDelayInbetween_1_0_3_2022-03-07_16-57.panel.zip)

I highly recommand to use the message as MemoryBlock to avoid errors in newer version of Ctrlr with sendMidiMessageNow :
`panel:sendMidiMessageNow(CtrlrMidiMessage(MemoryBlock({0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x04, 0x03, 0x00, 0x00, 0x00, tableModValue, 0xF7})))
`

**Reply by dnaldoog** [reply]
Hi Damien! It's a very nice example and helped my understanding of timers yet again. 🤢 I sometimes have trouble understanding them.

There are a couple of changes I would suggest though: 😁

<pre>
TimerCallback = function(timerId)
    if timerId == 0 then
        if counter1 < 16 then
            console("counter1 : " .. (counter1))

            local tableMod = {2, 6, 3, 18, 27, 4, 3, 6, 8, 6, 1, 5, 9, 15, 14, 21}

            for k, v in pairs(tableMod) do
                tableModValue = tableMod[counter1]
            end
            console("tablemod value : " .. tableModValue)
            panel:sendMidiMessageNow(
                CtrlrMidiMessage(
                    MemoryBlock(
                        {
                            0xF0,
                            0x0F,
                            0x03,
                            0x00,
                            0x11,
                            0x00,
                            0x00,
                            0x00,
                            0x00,
                            0x00,
                            0x01,
                            0x04,
                            0x03,
                            0x00,
                            0x00,
                            0x00,
                            tableModValue,
                            0xF7
                        }
                    )
                )
            )

            counter1 = counter1 + 1
        else
            timer:stopTimer(0)
        end
    end -- end timer #0
end -- f

</pre>

I think the loop (which should probably be ipairs not pairs) is unnecessary in that timer.

Also I think it's better to keep this table somewhere separate to the timer code otherwise the table is being recreated on each iteration of the timer.

<pre>
	     tableMod = {2,6,3,18,27,4,3,6,8,6,1,5,9,15,14,21} -- make this global in its own file
</pre>

or another suggestion for the Timer code:

<pre>
TimerCallback = function(timerId)
    if timerId == 0 then
        if counter1 == #tableMod then
            timer:stopTimer(timerId)
        else
            console("counter1 : " .. counter1)

            panel:sendMidiMessageNow(
                CtrlrMidiMessage(
                    MemoryBlock(
                        {
                            0xF0,
                            0x0F,
                            0x03,
                            0x00,
                            0x11,
                            0x00,
                            0x00,
                            0x00,
                            0x00,
                            0x00,
                            0x01,
                            0x04,
                            0x03,
                            0x00,
                            0x00,
                            0x00,
                            tableMod[counter1],
                            0xF7
                        }
                    )
                )
            )
            counter1 = counter1 + 1
        end
    end -- end timer #0
end -- f

tableMod = {2, 6, 3, 18, 27, 4, 3, 6, 8, 6, 1, 5, 9, 15, 14, 21} -- make this global in its own file

</pre>


Regards,

John.

----

Here's another take using your code
[Send Sysex With Delay Inbetween_2.zip](https://github.com/RomanKubiak/ctrlr/files/8210553/Send.Sysex.With.Delay.Inbetween_2.zip)

**Reply by DRBAUS** [reply]
Okay......I see how I could make the ENV Templates as tables that way and call them up with their own timer.

How do I go about changing the forelast byte in the sysex string though?
Some parameter values need to be 0x00 tableMod F7, some need to be 0x01 tableMod F7.

I would make another table, for example tableMod2, with values {0,1,1, 0 etc}. How to use the append Memoryblock in this case?

Played around a little with shortening the memoryblock in the constructor and

tmp:append(MemoryBlock(string.format("%2x %2x F7", tableMod2, tableMod[counter1]))) in the Callback but get a bad argument #2 to format, number expected, got table.

**Comment by dnaldoog** [comment]
Like this maybe?
<pre>
--
-- Called when the panel is created, no modulators will exist at this point
-- consider this the panels constructor
--
constructor = function()

	m=MemoryBlock("F0 0F 03 00 11 00 00 00 00 00 01 04 03 00 00") -- global
	tableMod = {2,6,3,18,27,4,3,6,8,6,1,5,9,15,14,21} -- global
	tab16th={0,1,0,1,0,0,1,0,0,1,1,1,0,1,0,0}


end

--
-- Called when a double-click event occurs on this component
--

myMethod = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)
	assert(#tab16th >= #tabMod)
	console("timer starts on double click")
	counter1 = 1
	console("timer starts on double click and Counter1 = 1")
	timer:setCallback(0, TimerCallback) -- TimerId 0
	timer:startTimer(0, 500) -- Timer Id 0 500ms


end


	
TimerCallback = function(timerId)
    if timerId == 0 then
        if counter1 == #tableMod then
            panel:getLabel("d"):setText("")
            timer:stopTimer(timerId)
        else
            panel:getLabel("d"):setText(counter1 .. "\187 ")

            local tmp = MemoryBlock(m)
	    tmp:append(MemoryBlock(string.format("%.2x",tab16th[counter1])))
            tmp:append(MemoryBlock(string.format("%.2x F7", tableMod[counter1])))
            panel:sendMidiMessageNow(CtrlrMidiMessage(tmp))
            counter1 = counter1 + 1
        end
    end -- end timer #0
end -- f
</pre>

----

[Send Sysex With Delay Inbetween_2_0_4.zip](https://github.com/RomanKubiak/ctrlr/files/8212117/Send.Sysex.With.Delay.Inbetween_2_0_4.zip)

**Reply by damiensellier** [reply]
This is just perfect! How to reduce  a 100 line script to 10. :)

**Reply by DRBAUS** [reply]
I have implemented this and it works great. I am looking for an additional tip now to reuse these timers for additional ENVs and possibly LAYERS.

The intention is to click on a label (ENV1, ENV2, ENV3) and depending on which label is clicked it should use a different MemoryBlock which is defined in the Constructor.

In the TimerCallback I have played around with something like:

sName= L(comp:getOwner():getName())

'if sName == ENV1 then local tmp = MemoryBlock(ENV1)
elseif sName == ENV2 then local tmp = MemoryBlock(ENV2)'

but since the text on the top of the function page is missing it does not recognize sName= L(comp:getOwner():getName())
When I put the text there the function does not work (of course?).

Where is a good place to put the 'if this label is clicked then use this MemoryBlock', 'elseif this label is clicked then use this Memoryblock'?

**Reply by dnaldoog** [reply]
Hi @DRBAUS - would have to see all the code (ie post the panel)  - it's difficult to get what you're asking, for me at least. :)

**Reply by DRBAUS** [reply]
Arigato for the quick response. 

Of course its better to post the panel also for you to see where I am heading in general with this.

**Reply by dnaldoog** [reply]
@DRBAUS  I downloaded it - will look at it later - you can delete it off the post, unless you want someone else to look at it!

**Reply by DRBAUS** [reply]
@dnaldoog Thanks. No rush.

**Reply by dnaldoog** [reply]
@DRBAUS 

Something like this?

Double click on each label to see a different sysex message appear in the __uiLabel__ below.

![DRBAUSImage](https://user-images.githubusercontent.com/76831683/159116004-9ac98ad5-1ac7-42f2-852a-d57f4476bbd3.png)



<pre>
constructor = function()

ENV1="F0 00 00 00 F7"
ENV2="F0 00 00 63 F7"
ENV3="F0 00 00 7F F7"
end -- function
---------------------------------------------------------------

fireTimer = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
-- attached to double click event on label
    local n = L(comp:getOwner():getName())
    local t = {ENV1 = 200, ENV2 = 260, ENV3 = 1200}

    timer:setCallback(t[n], TimerCallback) -- TimerId 0
    timer:startTimer(t[n], 1000) -- Timer Id + ms delay
end -- function
---------------------------------------------------------------
TimerCallback = function(timerId)
    if timerId == 200 then
        panel:getLabel("d"):setText(ENV1)
        timer:stopTimer(timerId)
    elseif timerId == 260 then
        panel:getLabel("d"):setText(ENV2)
        timer:stopTimer(timerId)
    elseif timerId == 1200 then
        panel:getLabel("d"):setText(ENV3)
        timer:stopTimer(timerId)
    end
end -- function
</pre>

I changed some of the code in your panel to make it faster. Please check to see if it is still working! :)

----

[ENSONIQ EPS-16 PLUS_1_72_2022-03-19_16-52.zip](https://github.com/RomanKubiak/ctrlr/files/8308915/ENSONIQ.EPS-16.PLUS_1_72_2022-03-19_16-52.zip)

[CLICK LABEL TIMER FIRE_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/8308917/CLICK.LABEL.TIMER.FIRE_1_0.zip)

**Reply by DRBAUS** [reply]
Thanks John. I am a bit overwhelmed here.

The Panel with the ENVs is not what I was looking for. There are ENV Presets that can be used for ENV1, ENV2 and ENV3. The modulator values (sysex) is the same for all modulators except for the byte that determines the ENV: 1, 2 or 3. I was looking to change the byte for the ENV. Last night I had the idea to use "when a mouse is over this component" to change the MemoryBlock that is being used and it works. (Just don't move your mouse over any of the other ENV display fields when the panel is sending the sysex to the hardware because it will change the memoryblock dynamically. There should be no reason to be doing anything else during the update process anyway).

On the 16Plus panel there are modulators that do not use consecutive sysex values. They skip values 02 and 07, something like that. Value 50 is not 0x32 but 0x01 0x00, etc.... and then it skips values again. That's why I made the script the way it was. Changing the last byte in the string to 'val' will not get the correct value in that case.

And....how did you so quickly get all the Modulator names in a script (AssignModulators)? I was going to do that but you beat me to it. Is there a code editor I need to check out or did you export the Modulator list?

Still looking at the other changes that you made and trying to figure out what they mean.

I apologize for my slow learning and I really appreciate the patience and help!

**Reply by dnaldoog** [reply]
Hi @DRBAUS 

before you retire for the night, can you point me towards the manual for this synthesizer?

I can’t access a computer right now, but will write later.

**Reply by DRBAUS** [reply]
The manual won't do.....but this will!
[Ensoniq_EPS16Plus_External_Command_Specification_130.zip](https://github.com/RomanKubiak/ctrlr/files/8312237/Ensoniq_EPS16Plus_External_Command_Specification_130.zip)

**Reply by dnaldoog** [reply]
> And....how did you so quickly get all the Modulator names in a script (AssignModulators)? I was going to do that but you beat me to it. Is there a code editor I need to check out or did you export the Modulator list?

Hi Robin,

There is a 'dev' function in that panel I sent.  Just run this code and then cut and paste the `console` output into the init function (when the panel loads). This creates lua variables, so a space and minus sign are illegal.

<pre>
for i=0,n-1 do
        mod = panel:getModulatorByIndex(i)
        local Name= L(mod:getName())
        Name=string.gsub(Name, "-",'_')
        Name=string.gsub(Name, " ",'_')
console(Name.."=panel:getModulatorByName(\""..Name.."\")")

--cut and paste output into your init script
		end

end
</pre>
 then instead of calling

`panel:getModulatorByName("ENV1"):getModulatorValue()`

you do `ENV1:getModulatorValue()`

**Reply by DRBAUS** [reply]
I am standing here with my weapon arsenal......and you beat me with a chopstick.....Sensei! 🗡️

**Reply by dnaldoog** [reply]
> Thanks John. I am a bit overwhelmed here.
> 
> The Panel with the ENVs is not what I was looking for. There are ENV Presets that can be used for ENV1, ENV2 and ENV3. The modulator values (sysex) is the same for all modulators except for the byte that determines the ENV: 1, 2 or 3. I was looking to change the byte for the ENV. Last night I had the idea to use "when a mouse is over this component" to change the MemoryBlock that is being used and it works. (Just don't move your mouse over any of the other ENV display fields when the panel is sending the sysex to the hardware because it will change the memoryblock dynamically. There should be no reason to be doing anything else during the update process anyway).

It sounds like you would need to use  some kind of global variable that is changed when you switch modulators or whatever control shares the same sysex and function. **Or** a local table could suffice.

----
<pre>
envSend = function(mod, value, source)
    if block then
        return
    end
    local n = L(mod:getName())
    local t = {ENV1 = 0x03, ENV2 = 0x05, ENV3 = 0x7f}
    local m = MemoryBlock(string.format("F0 0F 03 00 11 00 00 00 00 00 01 04 17 00 00 00 %.2X F7", t[n]))
    panel:sendMidiMessageNow(CtrlrMidiMessage(m:toHexString(1)))
end -- function
</pre>

**Reply by DRBAUS** [reply]
> Like this maybe?
> 
> --
> -- Called when the panel is created, no modulators will exist at this point
> -- consider this the panels constructor
> --
> constructor = function()
> 
> 	m=MemoryBlock("F0 0F 03 00 11 00 00 00 00 00 01 04 03 00 00") -- global
> 	tableMod = {2,6,3,18,27,4,3,6,8,6,1,5,9,15,14,21} -- global
> 	tab16th={0,1,0,1,0,0,1,0,0,1,1,1,0,1,0,0}
> 
> 
> end
> 
> --
> -- Called when a double-click event occurs on this component
> --
> 
> myMethod = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)
> 	assert(#tab16th >= #tabMod)
> 	console("timer starts on double click")
> 	counter1 = 1
> 	console("timer starts on double click and Counter1 = 1")
> 	timer:setCallback(0, TimerCallback) -- TimerId 0
> 	timer:startTimer(0, 500) -- Timer Id 0 500ms
> 
> 
> end
> 
> 
> 	
> TimerCallback = function(timerId)
>     if timerId == 0 then
>         if counter1 == #tableMod then
>             panel:getLabel("d"):setText("")
>             timer:stopTimer(timerId)
>         else
>             panel:getLabel("d"):setText(counter1 .. "\187 ")
> 
>             local tmp = MemoryBlock(m)
> 	    tmp:append(MemoryBlock(string.format("%.2x",tab16th[counter1])))
>             tmp:append(MemoryBlock(string.format("%.2x F7", tableMod[counter1])))
>             panel:sendMidiMessageNow(CtrlrMidiMessage(tmp))
>             counter1 = counter1 + 1
>         end
>     end -- end timer #0
> end -- f
> 
> [Send Sysex With Delay Inbetween_2_0_4.zip](https://github.com/RomanKubiak/ctrlr/files/8212117/Send.Sysex.With.Delay.Inbetween_2_0_4.zip)

This code works well. However I would like to 're-use' the timers for all 8 Layers and all 8 Wavesamples and somehow want to manipulate MemoryBlock in the Constructor but for days I have been looking to come up with something without any working result.

m=MemoryBlock("F0 0F 03 00 11 00 00 00 **00** 00 **01** **04** 03 00 00") -- global

How can I change these values if i.e. one Modulator is 6 (Layer 6), another Modulator is 3 (Wavesample 3) and the last Modulator is 2 (Envelope 2).

Inserting sName etc won't work in this function somehow.

**Reply by dnaldoog** [reply]
Hi Baus!

You can always have an array of `MemoryBlock` objects and assign to each layer - ie access the array number for each layer `marray[6]` for layer 6 etc.

You could then write to each MemoryBlock as required `marray[3]:setByte(11,127)` etc That value of 127 might come from a `uiSlider` for example.
<pre>
    str = "F0 0F 03 00 11 00 00 00 00 00"
    marray = {}
    for i = 1, 8 do
        marray[i] = MemoryBlock()
        marray[i]:loadFromHexString(str)
        marray[i]:append(MemoryBlock(string.format("%.2x 04 03 00 00", i)))
    end
</pre>

----
[array of memoryblock_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/8610611/array.of.memoryblock_1_0.zip)

**Reply by DRBAUS** [reply]
Thank you dnaldoog!

I understand how that works but fail to understand how to apply it correctly. I will look into it more in-depth the next few days.

**Reply by DRBAUS** [reply]
I don't get!

I mean....how do you append a string without changing again the values before the append?

marray[i]:append(MemoryBlock(string.format("%.2x 04 03 00 00", i))) The modulator changes the %.2x 

How do I change the 04 then with another modulator without changing the byte before it?

**Reply by dnaldoog** [reply]
> I don't get!
> 
> I mean....how do you append a string without changing again the values before the append?
> 
> marray[i]:append(MemoryBlock(.format("%.2x 04 03 00 00", i))) The modulator changes the %.2x
> 
> How do I change the 04 then with string another modulator without changing the byte before it?
 
 just on train so this will be short.


`var2=45`
<pre>
marray[i]:append(MemoryBlock(.format("%.2x %.2x 03 00 00", i,var2))) 
</pre>

**Reply by DRBAUS** [reply]
Okay....with this I have been able to create what I need.

I think I need to get the tables out of the CONSTRUCTOR function in my panel. Constructor is loaded before any modulators are created so having a modulator change the byte won't work? Is that correct?

When I create a function and use it on modulator change is looks like it's working.

**Reply by dnaldoog** [reply]
> Okay....with this I have been able to create what I need.
> 
> I think I need to get the tables out of the CONSTRUCTOR function in my panel. Constructor is loaded before any modulators are created so having a modulator change the byte won't work? Is that correct?

That is correct. 🙏

**Reply by DRBAUS** [reply]
I am stuck and I don't know how to explain in writing either what I mean.

Would it help to make a video explaining what I want?

**Reply by DRBAUS** [reply]
Is it possible to append a sysex string from 2 different functions?

Sorry, I am frustrated a bit.

**Reply by dnaldoog** [reply]
> Is it possible to append a sysex string from 2 different functions?
> 
> Sorry, I am frustrated a bit.

Yes, if it’s a global string any function can access it, but you have to understand global variables vs local in lua and scope in programming and more importantly be able to keep track of those variables in your program. If you don’t keep track of them, then you’ll end up with a global soup of variables and your program could become very unstable. Are you well familiar with all that?

**Reply by dnaldoog** [reply]
> I am stuck and I don't know how to explain in writing either what I mean.
> 
> Would it help to make a video explaining what I want?

Perhaps???

**Reply by DRBAUS** [reply]
Wow....that was easier than I thought...literally!

I took all the tables out of the Constructor and placed them in the AssignModulators function which is called when the panel has finished loading.

Since it's just a MemoryBlock that I want to change it only took some simple 'elseif then' statements to change that whole MemoryBlock with the string that I need. The other code that will append the string is already in place so....it's working!

Thanks for putting up with me!

**Reply by DRBAUS** [reply]
> > And....how did you so quickly get all the Modulator names in a script (AssignModulators)? I was going to do that but you beat me to it. Is there a code editor I need to check out or did you export the Modulator list?
> 
> Hi Robin,
> 
> There is a 'dev' function in that panel I sent. Just run this code and then cut and paste the `console` output into the init function (when the panel loads). This creates lua variables, so a space and minus sign are illegal.
> 
> for i=0,n-1 do
>         mod = panel:getModulatorByIndex(i)
>         local Name= L(mod:getName())
>         Name=string.gsub(Name, "-",'_')
>         Name=string.gsub(Name, " ",'_')
> console(Name.."=panel:getModulatorByName(\""..Name.."\")")
> 
> --cut and paste output into your init script
> 		end
> 
> end
> 
> then instead of calling
> 
> `panel:getModulatorByName("ENV1"):getModulatorValue()`
> 
> you do `ENV1:getModulatorValue()`

What could be the cause that all of a sudden this script is not working anymore? Clicking the dev button does not output anything anymore and I did not change anything in the scripts.

**Reply by dnaldoog** [reply]
Hi @DRBAUS 

You’d need to post the panel for me to answer that ☝️ one.

**Comment by DRBAUS** [comment]
Great!!!

I see now how I can manipulate that sysex string.

On to make the ENV TEMPLATES and put them to work.

Thanks everyone!

Op wo 9 mrt. 2022 07:08 schreef John Goodland ***@***.***>:

**Reply by damiensellier** [reply]
@DRBAUS can you please select John's @dnaldoog previous post and mark it as the answer? thanks
